app/src/org/commcare/dalvik/activities/UpdateActivity.java

```diff
@@ -0,0 +1,256 @@
+ * - Stop a downloading update
+ * - Apply a downloaded update
+ *
+ * @author Phillip Mates (pmates@dimagi.com)
+ */
+public class UpdateActivity extends CommCareActivity<UpdateActivity>
+        implements TaskListener<Integer, AppInstallStatus> {
+
+    private static final String TAG = UpdateActivity.class.getSimpleName();
+    private static final String TASK_CANCELLING_KEY = "update_task_cancelling";
+    private static final int DIALOG_UPGRADE_INSTALL = 6;
+
+    private boolean taskIsCancelling;
+    private UpdateTask updateTask;
+    private UpdateUiController uiController;
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+
+        uiController = new UpdateUiController(this);
+
+        loadSaveInstanceState(savedInstanceState);
+
+        setupUpdateTask();
+    }
+
+    private void loadSaveInstanceState(Bundle savedInstanceState) {
+        if (savedInstanceState != null) {
+            taskIsCancelling =
+                    savedInstanceState.getBoolean(TASK_CANCELLING_KEY, false);
+        }
+    }
+
+    private void setupUpdateTask() {
+        updateTask = UpdateTask.getRunningInstance();
+
+        try {
+            if (updateTask != null) {
+                updateTask.registerTaskListener(this);
+            }
+        } catch (TaskListenerRegistrationException e) {
+            Log.e(TAG, "Attempting to register a TaskListener to an already " +
+                    "registered task.");
+            uiController.errorUiState();
+        }
+    }
+
+    @Override
+    protected void onResume() {
+        super.onResume();
+
+        if (ConnectivityStatus.isNetworkNotConnected(this) &&
+                ConnectivityStatus.isAirplaneModeOn(this)) {
+            uiController.noConnectivityUiState();
+            return;
+        }
+
+        int currentProgress = 0;
+        int maxProgress = 0;
+        if (updateTask != null) {
+            currentProgress = updateTask.getProgress();
+            maxProgress = updateTask.getMaxProgress();
+            if (taskIsCancelling) {
+                uiController.cancellingUiState();
+            } else {
+                setUiStateFromRunningTask(updateTask.getStatus());
+            }
+        } else {
+            pendingUpdateOrIdle();
+        }
+        uiController.updateProgressBar(currentProgress, maxProgress);
+        uiController.refreshStatusText();
+    }
+
+    private void setUiStateFromRunningTask(AsyncTask.Status taskStatus) {
+        switch (taskStatus) {
+            case RUNNING:
+                uiController.downloadingUiState();
+                break;
+            case PENDING:
+                pendingUpdateOrIdle();
+                break;
+            case FINISHED:
+                uiController.errorUiState();
+                break;
+            default:
+                uiController.errorUiState();
+        }
+    }
+
+    private void pendingUpdateOrIdle() {
```

Comment:
Wanted to have all the decision logic be in one place, mainly the activity. I should probably rename the `UpdateUiController` to `UpdateUiState` to signify that it is just used to enter different UI states, but doesn't actually contain any decision logic.